All auth changes

distantnative opened the pull request:
## Description
<!-- 
Add info about why this PR exists and the decisions that went into it.
This info is meant for the reviewer of this PR.
 
You may keep it short or omit it if it's a simple PR. Please add more
context and a summary of changes if it's a more complex PR. 

Make sure to point your PR to the relevant develop branches, e.g.
`develop-patch`, `develop-minor` or `v6/develop`.

How to contribute: https://contribute.getkirby.com
-->

- [ ] https://github.com/getkirby/kirby/pull/8045
- [ ] Decide where to add `@internal`
- [ ] Hardening
  - [ ] `Limits` to use new `F::modify()` (read-and-write under lock)
  - [ ] Hash email in auth limits log
  - [ ] Per-challenge max attempts

## Changelog 
<!--
Add relevant release notes. Keep the target audience (Kirby user) in mind.
Reference issues from the `kirby` repo  or ideas from `feedback.getkirby.com`.
-->

### ✨ Enhancements
<!-- 
e.g. Improve a11y of feature X
-->
- New `authMethods` plugin extension
- New unified security drawer for users to manage all auth-related settings
    <img width="948" height="1120" alt="image" src="https://github.com/user-attachments/assets/6a46335c-a09e-471e-bc92-b1eac92a71f6" />
- New `hashtag` and `email-unread` icons

### ♻️ Refactored
<!-- 
e.g. Rename method X to method Y.
-->
- New `Kirby\Auth` package:
  - New `Kirby\Auth\Limits`, `Kirby\Auth\Csrf` classes
  - New `Kirby\Auth\Exception\RateLimitException`  exception
  - New `Kirby\Auth\Methods` and individual `Kirby\Auth\Method` classes
  - New `Kirby\Auth\Exception\LoginNotPermittedException`
  - New `Kirby\Auth\Method\BasicAuthMethod` class
  - New `Kirby\Auth\User`
  - New `Kirby\Auth\Challenge` abstract class and classes for the two core challenges
  - New `Kirby\Auth\Challenges`
  - New `Kirby\Auth\Pending` class for storing data of a pending challenge
  - New `Kirby\Exception\UserNotFoundException` and `Kirby\Auth\Exception\ChallengeTimeoutException`
  - New`Kirby\Auth\Status` class and `Kirby\Auth\State` enum
  - New `$user->changeSecret($type, $secret)` method
- User settings dropdown only has a `Security` entry now
- New `UserTotpDrawerController` and `<k-user-totp-drawer>` to manage TOTP challenge settings
  <img width="427" height="689" alt="Screenshot 2026-02-03 at 16 05 45" src="https://github.com/user-attachments/assets/016a0c33-01b7-4703-ac16-6945f5cced45" />


### ☠️ Deprecated
<!-- 
e.g. Deprecate method X. Use method Y instead.
-->
- The following methods of `Kirby\Cms\Auth` have been deprecated: 
  - `::checkRateLimit()`, `::isBlocked()`, `::log()`, `::logfile()` and `::track()` → Use `$auth->limits()` and the respective method instead.
  - `Kirby\Cms\Auth::login2fa()` → Use `Kirby\Cms\Auth::authenticate()` instead
  - `Kirby\Cms\System::loginMethods()`  → Use `$kirby->auth()->methods()->enabled()` instead
  - `Kirby\Cms\Auth\Challenge`, `Kirby\Cms\Auth\EmailChallenge`, `Kirby\Cms\Auth\TotpChallenge`
  - `$auth->enabledChallenges()` → Use `$auth->challenges()->enabled()` instead
  - `Kirby\Cms\Auth\Status` → Use `Kirby\Auth\Status` instead
  - `Kirby\Cms\Auth` has been renamed to `Kirby\Auth\Auth`. A class alias has been added but will be removed in a future major release.
  - `$user->changeTotp()` has been deprecated → Use `$user->changeSecret('totp', $secret)` instead

### 🧹 Housekeeping
<!-- 
e.g. Update JS dependencies
-->

### 🚨 Breaking changes
<!-- 
e.g. Method X has been removed
-->

- Everywhere where `Kirby\Cms\Auth\Status` got returned or expected as parameter, `Kirby\Auth\Status` is now used.
- Removed `UserTotpEnableDialogController`, `UserTotpDisableDialogController` and `<k-totp-dialog>`. Use `UserTotpDrawerController ` and `<k-user-totp-drawer>` instead.


## Documentation

```php
public class FooMethod extends \Kirby\Auth\Method
{
	public function authenticate(
		string $email,
		#[SensitiveParameter]
		string|null $password = null,
		bool $long = false
	): User|Status {
		// returns either the logged in user
		// or creates a challenge and returns the 
		// pending auth status
	}

	public static function isAvailable(
		Auth $auth,
		array $options = []
	): bool {
		// some condition if the auth method
		// should be available, by default true
	}

	public static function isUsingChallenges(
		Auth $auth,
		array $options = []
	): bool {
		// if a challenge is created by the method,
		// default false
	}
}

// register custom auth method
App::plugin(
	name: 'my/plugin',
	extensions: ['authMethods' => ['foo' => FooMethod::class]]
);
```


```php
class FooChallenge extends \Kirby\Auth\Challenge
{
	public function create(): Pending|null
	{
		// create the challenge and return a Pending object
		// that can store secret as well as public (for the frontend)
		// data to later be used in ::verify().
		// If no data is needed, return null.

		return new Pending(
			secret: 'my-secret-hash',
			public: ['foo' => 'bar']
		);
	}

	public static function isAvailable(User $user, string $mode): bool
	{
		// check if the challenge is available for the user,
		// e.g. the user has set it up
	}

	public function verify(
		#[SensitiveParameter]
		mixed $input,
		Pending $data
	): bool {
		// receives the input from the frontend as well as
		// the Pending data from ::create() and should
		// evaluate the challenge
	}
}

// register the custom challenge
App::plugin(
	name: 'my/plugin',
	extensions: ['authChallenges' => ['foo' => FooChallenge::class]]
);
```


### For review team
<!--
We will take care of the following before merging the PR.
-->

- [ ] Add changes & docs to release notes draft in Notion

distantnative committed:
refact: `Kirby\Auth\Csrf`

distantnative committed:
refact: `Kirby\Auth\Limits`

distantnative committed:
refact: `Kirby\Auth\Methods`

distantnative committed:
docs: Add missing docblock

distantnative committed:
refact: `Kirby\Auth\Method\BasicAuthMethod`

distantnative committed:
refact: `Kirby\Auth\Method\BasicAuthMethod`

distantnative committed:
refact: `Method::isAvailable()`

distantnative committed:
refact: `Method::isUsingChallenges()`

distantnative committed:
refact: `Kirby\Auth\User`

distantnative committed:
refact: Keep using `:: currentUserFromBasicAuth()`

distantnative committed:
refact: Keep using `:: currentUserFromBasicAuth()`

distantnative committed:
refact: Remove internal deprec `:: checkRateLimit`

distantnative committed:
refact: `Kirby\Auth\State`

distantnative committed:
refact: `Kirby\Auth\Status`

distantnative committed:
refact: Migrate `Auth` class to `Kirby\Auth\Auth`

distantnative committed:
refact: `Kirby\Auth\Challenges`

distantnative committed:
fix: Use `LoginNotPermittedException`

distantnative committed:
refact: Move `UserNotFoundException`

distantnative committed:
test: Add more unit tests

distantnative committed:
fix: Pass full status array to login view

distantnative committed:
refact: `$user->changeSecret()`

distantnative committed:
test: Add more unit tests

distantnative committed:
fix: `Method::get()` is config-agnostic

distantnative committed:
test: Use auth test case

distantnative committed:
test: Use auth test case

distantnative committed:
style: Fix cs

distantnative committed:
test: Use auth test case

distantnative committed:
test: Sort tests alphabetically

distantnative committed:
test: Test if session keys are created

distantnative committed:
test: Add missing previous tests

distantnative committed:
refact: `::isAvailable()` always throws

bastianallgeier committed:
Merge pull request #7844 from getkirby/v6/refact/auth-methods

[Auth 4] refact: `Kirby\Auth\Methods`

bastianallgeier committed:
Merge pull request #7845 from getkirby/v6/refact/auth-basic-method

[Auth 5] refact: `Kirby\Auth\Method\BasicAuthMethod`

bastianallgeier committed:
Merge branch 'v6/refact/auth' into v6/refact/auth-user

bastianallgeier committed:
Merge pull request #7847 from getkirby/v6/refact/auth-user

[Auth 6] refact: `Kirby\Auth\User`

distantnative committed:
refct: Use `User::hashPassword()`

distantnative committed:
refact: Move exception to `Challenges::create()`

distantnative committed:
fix: Set full session info in `Challenges::create`

distantnative committed:
refact: `Challenge::isEnabled()`

distantnative committed:
refact: `Methods::isEnabled()`

distantnative committed:
test: Add missing unit tests

distantnative committed:
style: Fix cs

bastianallgeier committed:
Merge pull request #7849 from getkirby/v6/refact/auth-challenges

[Auth 7] refact: `Kirby\Auth\Challenges`

bastianallgeier committed:
Merge pull request #7850 from getkirby/v6/refact/auth-status-state

[Auth 8] refact: `Kirby\Auth\Status` + `Kirby\Auth\State`

bastianallgeier committed:
Merge pull request #7856 from getkirby/v6/refact/auth-class-migration

[Auth 9] refact: Migrate `Auth` class to `Kirby\Auth\Auth`

distantnative committed:
fix: User rule type

bastianallgeier committed:
Merge pull request #7853 from getkirby/v6/refact/auth-change-secret

[Auth] refact: `$user->changeSecret()`

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
test: Fix unit test performance

lukasbestle commented on src/Cms/UserActions.php:
```diff
@@ -114,23 +114,41 @@ public function changeRole(string $role): static
+	 * Stores secrets for the user
+	 * @since 6.0.0
+	 */
+	public function changeSecret(string $secret, mixed $content): static
```
Comment for merged PR #7853:

`Suggestion` to ensure that the secret isn't printed in logs (like with the old `changeTotp()`):

```suggestion
	public function changeSecret(
		string $secret,
		#[SensitiveParameter]
		mixed $content
	): static
```

lukasbestle commented on src/Cms/UserRules.php:
```diff
@@ -133,9 +133,34 @@ public static function changeRole(User $user, string $role): void
+			throw new PermissionException(
+				message: 'You cannot change user secrets for ' . $user->email()
+			);
+		}
```
`Suggestion`: Can we add the the "wrong length" check from `changeTotp` here if `$secret === 'totp'`?

distantnative replied:
You mean that `UserRules::changeTotp()` runs `new Totp($secret)`? I thought about it when working on the PR but actually think this does not belong in the `::changeSecret()` method. This one should be concerned with writing any content as secret to the user's file. Whether the content is valid for whatever use case should be the concern of the code that uses `::changeSecret()` to update it. This already happens for `UserTotpEnableDialogController::submit()` which is where we update the TOTP secret.

lukasbestle replied:
Good point. So far the additional check in `changeTotp` was an extra safety net, but it doesn't really scale to add these for every kind of secret.

distantnative committed:
fix: `$content` as sensitive in `UserActions::changeSecret()`

Co-authored-by: Lukas Bestle <lukas@getkirby.com>

lukasbestle commented on src/Auth/Limits.php:
```diff
@@ -0,0 +1,114 @@
+		$ip     = $this->kirby->visitor()->ip(hash: true);
+		$trials = $this->kirby->option('auth.trials', 10);
+
+		if (($log['by-ip'][$ip]['trials'] ?? null) >= $trials) {
```
`Nitpick`: `null >= 10` is currently interpreted as `false` by PHP, but this would be more explicit:


```suggestion
		if (($log['by-ip'][$ip]['trials'] ?? 0) >= $trials) {
```

lukasbestle commented on src/Auth/Limits.php:
```diff
@@ -0,0 +1,114 @@
+		}
+
+		if ($this->kirby->users()->find($email)) {
+			if (($log['by-email'][$email]['trials'] ?? null) >= $trials) {
```
```suggestion
			if (($log['by-email'][$email]['trials'] ?? 0) >= $trials) {
```

lukasbestle commented on src/Auth/Auth.php:
```diff
@@ -0,0 +1,612 @@
+
+	/**
+	 * Check if logins are blocked for the current ip or email
+	 * @deprecated 6.0.0 Use `self::limits()->Blocked()` instead
```
`Typo fix`:

```suggestion
	 * @deprecated 6.0.0 Use `self::limits()->isBlocked()` instead
```

lukasbestle commented on src/Auth/Auth.php:
```diff
@@ -0,0 +1,612 @@
+			if ($e instanceof RateLimitException === false) {
+				try {
+					$this->limits->track($email);
+				} catch (Throwable) {
```
`Issue`: Unrelated to this PR series, but I just noticed that we accidentally removed the variable override in https://github.com/getkirby/kirby/commit/c0b60ddac9312f8499b0af288f435678c82b6478:


```suggestion
				} catch (Throwable $e) {
```

Maybe there is a better way to do this that doesn't look "unused" to automated tools.

distantnative replied:
Good catch! Even with the comment, this went totally over my head to fully follow the code path here. Maybe to make this more explicit, we should add a variable to really keep track of it. E.g. on line 520 `$error = $e` and then on line 525 override it again. That way, I think automated tools/IDEs would also not flag it as unused.

distantnative replied:
@lukasbestle Implemented it via `$error = $e`

lukasbestle reviewed (commented):
Comments for PR 1: #7835

lukasbestle committed:
style: Fix CS

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
fix: Apply suggestions from code review

Co-authored-by: Lukas Bestle <lukas@getkirby.com>

distantnative committed:
fix: Overwriting error in `::validatePassword()`

distantnative committed:
test: Use `Kirby\TestCase` for auth exceptions

distantnative committed:
feat: User security drawer

distantnative committed:
refact!: User security TOTP drawer

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
test: Fix user totp drawer controller unit test

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
fix: Solve Psalm reported inconsistencies

distantnative committed:
fix: Some auth refactoring fixes after review

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
[Auth] refact!: Remove legacy auth challenges support (#8044)

refact!: Remove legacy auth challenges support

BREAKING CHANGE: Custom auth challenges extending `Kirby\Cms\Auth\Challenge` have to be adapted to use `Kirby\Auth\Challenge`

distantnative committed:
[Auth] feat: Switch between auth challenges (#8063)

feat: Switch between auth challenges

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
[Auth] feat: New login Vue components (#8077)

feat: New login Vue components

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
[Auth] refact: `::form()` methods for methods/challenges (#8122)

refact: `::form()` methods for methods/challenges

distantnative committed:
[Auth] refact: Allow auth methods without email (#8123)

refact: Allow auth methods without email

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
fix: Regression in `::authenticateApiRequest()`

distantnative committed:
fix: Do not fire user login failed hook twice

distantnative committed:
fix: `BasicAuth::isAttempted()`

distantnative committed:
fix: Auth user password timestamp check

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

distantnative committed:
Merge branch 'v6/develop' into v6/refact/auth

bastianallgeier commented on src/Auth/Auth.php:
```diff
@@ -0,0 +1,617 @@
+		if ($session === null && $allowImpersonation === true) {
+			$this->status = $status;
+		}
+
```
I think the `$session === null` check can never really happen here, because of `$session = $this->session($session);` in line 420

bastianallgeier commented on src/Auth/Auth.php:
```diff
@@ -0,0 +1,617 @@
+	/**
+	 * Become any existing user or disable the current user
+	 *
+	 * @param 'string'|null $who User ID or email address,
```
```suggestion
	 * @param string|null $who User ID or email address,
```

bastianallgeier commented on src/Auth/Method/PasswordResetMethod.php:
```diff
@@ -0,0 +1,59 @@
+{
+	public function authenticate(
+		string|null $email,
+		string|null $password = null,
```
Could be better to repeat it here. 

```suggestion
		string|null $password = null, #[SensitiveParameter]
```

distantnative replied:
I think it would need to go on the line before $password. I left it initially out as no password is passed here. But you're right, better for consistency to always have it in any `*Method::authenticate()` signature.

bastianallgeier commented on src/Panel/Controller/View/LoginViewController.php:
```diff
@@ -12,40 +21,154 @@
+
+		$challenges = $this->kirby->auth()->challenges();
+		$email      = $this->status->email();
+		$user       = $this->kirby->user($email);
```
`$this->kirby->user($email)` could return null if the user may have been deleted after the challenge was created. Unlikely but possible. Do we want to catch that and redirect to the login in this case?

distantnative replied:
This one is tricky cause the case is not just for a user that is deleted after the challenge was created. A challenge is actually created for any email address, no matter if existing user or not - to avoid enumerating on potentially existing email addresses. So submitting a valid vs. invalid email must look identical. I'll look for a fix that upholds that.

distantnative replied:
I think I have found a way: https://github.com/getkirby/kirby/pull/7848/commits/2f0c80578b74a8b388dd09df9be48a3feebe3cf7

bastianallgeier replied:
Ah, that's a great fix.
